Source/Core/Core/PowerPC/Jit64/Jit.cpp

```diff
@@ -785,12 +804,67 @@ void Jit64::Jit(u32 em_address)
     return;
   }
 
-  JitBlock* b = blocks.AllocateBlock(em_address);
-  DoJit(em_address, b, nextPC);
-  blocks.FinalizeBlock(*b, jo.enableBlocklink, code_block.m_physical_addresses);
+  if (SetEmitterStateToFreeCodeRegion())
+  {
+    u8* near_start = GetWritableCodePtr();
+    u8* far_start = m_far_code.GetWritableCodePtr();
+
+    JitBlock* b = blocks.AllocateBlock(em_address);
+    if (DoJit(em_address, b, nextPC))
+    {
+      // Code generation succeeded.
+
+      // Mark the memory regions that this code block uses as used in the local rangesets.
+      u8* near_end = GetWritableCodePtr();
+      if (near_start != near_end)
+        m_free_ranges_near.erase(near_start, near_end);
+      u8* far_end = m_far_code.GetWritableCodePtr();
+      if (far_start != far_end)
+        m_free_ranges_far.erase(far_start, far_end);
+
+      // Store the used memory regions in the block so we know what to mark as unused when the
+      // block gets invalidated.
+      b->near_begin = near_start;
+      b->near_end = near_end;
+      b->far_begin = far_start;
+      b->far_end = far_end;
+
+      blocks.FinalizeBlock(*b, jo.enableBlocklink, code_block.m_physical_addresses);
+      return;
+    }
+  }
+
+  // Code generation failed due to not enough free space in either the near or far code regions.
+  // Clear the entire JIT cache and retry.
+  WARN_LOG(POWERPC, "flushing code caches, please report if this happens a lot");
+  ClearCache();
+  Jit(em_address);
```

Comment:
Nothing keeps us from running into a potential stack overflow here (by calling `Jit` over and over again after flushing); can this be a potential issue in practise? Or is it unlikely that we ever emit so much code for a block that it ever fills up everything available?

Reply:
Considering the JIT previously would just start writing into arbitrary memory if the JITed code was larger than the remaining buffer length and we've never -- at least knowingly -- ran into it, I'd say this is impossible in practice unless you designed PowerPC code to specifically exploit Dolphin's JIT. I guess a check is easy enough to add though, so might as well.

Reply:
I added a PanicAlert and exit(), is that okay?

Reply:
Seems ok to me, thanks!